image.go

```diff
@@ -358,64 +326,409 @@ func (i *containerImageRef) createConfigsAndManifests() (v1.Image, v1.Manifest,
+	// matching fields in both formats.
+	if err := config.OverrideOCI(&oimage.Config, i.overrideChanges, i.overrideConfig); err != nil {
+		return nil, fmt.Errorf("applying changes: %w", err)
+	}
+
+	// If we're producing a confidential workload, override the command and
+	// assorted other settings that aren't expected to work correctly.
+	if i.confidentialWorkload.Convert {
+		oimage.Config.Entrypoint = []string{"/entrypoint"}
+		oimage.Config.Cmd = nil
+		oimage.Config.User = ""
+		oimage.Config.WorkingDir = ""
+		oimage.Config.Volumes = nil
+		oimage.Config.ExposedPorts = nil
+	}
+
+	// Build empty manifest.  The Layers lists will be populated later.
+	return &ociManifestBuilder{
+		i: i,
+		// The default layer media type assumes no compression.
+		mediaType: v1.MediaTypeImageLayer,
+		oimage:    oimage,
+		omanifest: v1.Manifest{
+			Versioned: specs.Versioned{
+				SchemaVersion: 2,
+			},
+			MediaType: v1.MediaTypeImageManifest,
+			Config: v1.Descriptor{
+				MediaType: v1.MediaTypeImageConfig,
+			},
+			Layers:      []v1.Descriptor{},
+			Annotations: i.annotations,
 		},
-		Layers: []docker.V2S2Descriptor{},
+	}, nil
+}
+
+func (mb *ociManifestBuilder) addLayer(layerBlobSum digest.Digest, layerBlobSize int64, diffID digest.Digest) {
+	olayerDescriptor := v1.Descriptor{
+		MediaType: mb.mediaType,
+		Digest:    layerBlobSum,
+		Size:      layerBlobSize,
 	}
+	mb.omanifest.Layers = append(mb.omanifest.Layers, olayerDescriptor)
+	// Note this layer in the list of diffIDs, again using the uncompressed digest.
+	mb.oimage.RootFS.DiffIDs = append(mb.oimage.RootFS.DiffIDs, diffID)
+}
 
-	return oimage, omanifest, dimage, dmanifest, nil
+// Compute the media types which we need to attach to a layer, given the type of
+// compression that we'll be applying.
+func (mb *ociManifestBuilder) computeLayerMIMEType(what string, layerCompression archive.Compression) error {
+	omediaType := v1.MediaTypeImageLayer
+	if layerCompression != archive.Uncompressed {
+		switch layerCompression {
+		case archive.Gzip:
+			omediaType = v1.MediaTypeImageLayerGzip
+			logrus.Debugf("compressing %s with gzip", what)
+		case archive.Bzip2:
+			// Until the image specs define a media type for bzip2-compressed layers, even if we know
+			// how to decompress them, we can't try to compress layers with bzip2.
+			return errors.New("media type for bzip2-compressed layers is not defined")
+		case archive.Xz:
+			// Until the image specs define a media type for xz-compressed layers, even if we know
+			// how to decompress them, we can't try to compress layers with xz.
+			return errors.New("media type for xz-compressed layers is not defined")
+		case archive.Zstd:
+			// Until the image specs define a media type for zstd-compressed layers, even if we know
+			// how to decompress them, we can't try to compress layers with zstd.
+			return errors.New("media type for zstd-compressed layers is not defined")
```

Comment:
This could set `omediaType` to `v1.MediaTypeImageLayerZstd` instead of returning an error.

Reply:
In #5452 I was asked not to support zstd in Docker manifests because the spec doesn't officially support it. That was the impetus for this refactor.

Reply:
Oh, sorry, this is for the OCI manifest. The idea was to make this change in #5452 once this PR is merged, but I could unify the two PRs if you prefer.